File: engine/src/main/java/de/hanno/hpengine/engine/threads/TimeStepThread.java
package de.hanno.hpengine.engine.threads;

import java.util.concurrent.TimeUnit;
import java.util.logging.LogManager;
import java.util.logging.Logger;

public abstract class TimeStepThread extends Thread {
    protected final String name;
    private Logger LOGGER;

    private long start = 0l;
    protected long lastFrame = 0l;

    public boolean stopRequested = false;
    protected volatile float minimumCycleTimeInSeconds;

    //TODO: Consider this public
    protected TimeStepThread(String name) {
        this(name, 0.0f);
    }
    public TimeStepThread(String name, float minimumCycleTimeInSeconds) {
        super();
        LOGGER = LogManager.getLogManager().getLogger(TimeStepThread.class.getName() + " " + name);
        setMinimumCycleTimeInSeconds(minimumCycleTimeInSeconds);
        this.name = name;
        setUncaughtExceptionHandler((t, e) -> {
            LOGGER.severe("An error!");
            e.printStackTrace();
        });
    }

    @Override
    public void start() {
        setDaemon(true);
        super.start();
        start = System.nanoTime();
        lastFrame = System.nanoTime();
    }
    @Override
    public void run() {
        setName(name);

        while(!stopRequested) {
            long ns = System.nanoTime() - lastFrame;

            float seconds = TimeUnit.NANOSECONDS.toSeconds(ns);
            update(seconds);
            waitIfNecessary(seconds);
            lastFrame = System.nanoTime();
        }
        cleanup();
    }

    protected void waitIfNecessary(float actualS) {
        float secondsLeft = (getMinimumCycleTimeInSeconds() - actualS);
        if(secondsLeft <= 0) { return; }

        while(secondsLeft >= 0) {
            try {
                long timeBeforeSleep = System.nanoTime();
                Thread.sleep(0, 100000);
                long sleptNanoSeconds = System.nanoTime() - timeBeforeSleep;
                double sleptSeconds = sleptNanoSeconds / 1000000000.0;
                secondsLeft -= sleptSeconds;
            } catch (InterruptedException e) {
                e.printStackTrace();
            }

        }
    }

    public void cleanup() {

    }

    public abstract void update(float seconds);

    private void setMinimumCycleTimeInSeconds(float minimumCycleTimeInSeconds) {
        this.minimumCycleTimeInSeconds = minimumCycleTimeInSeconds;
    }

    public float getMinimumCycleTimeInSeconds() {
        return minimumCycleTimeInSeconds;
    }
}

File: engine/src/main/kotlin/de/hanno/hpengine/engine/threads/TimeStepThread.kt
package de.hanno.hpengine.engine.threads

import java.lang.InterruptedException
import de.hanno.hpengine.engine.threads.TimeStepThread
import java.util.concurrent.TimeUnit
import java.util.logging.LogManager
import java.util.logging.Logger

abstract class TimeStepThread(name: String, val minimumCycleTimeInSeconds: Float) : Thread() {
    private var start = 0L
    protected var lastFrame = 0L
    var stopRequested = false

    init {
        this.name = name
        uncaughtExceptionHandler = UncaughtExceptionHandler { _: Thread?, e: Throwable ->
            e.printStackTrace()
        }
    }

    //TODO: Consider this public
    protected constructor(name: String) : this(name, 0.0f) {}

    override fun start() {
        isDaemon = true
        super.start()
        start = System.nanoTime()
        lastFrame = System.nanoTime()
    }

    override fun run() {
        setName(name)
        while (!stopRequested) {
            val ns = System.nanoTime() - lastFrame
            val seconds = TimeUnit.NANOSECONDS.toSeconds(ns).toFloat()
            update(seconds)
            waitIfNecessary(seconds)
            lastFrame = System.nanoTime()
        }
        cleanup()
    }

    protected fun waitIfNecessary(actualS: Float) {
        var secondsLeft = minimumCycleTimeInSeconds - actualS
        if (secondsLeft <= 0) {
            return
        }
        while (secondsLeft >= 0) {
            try {
                val timeBeforeSleep = System.nanoTime()
                sleep(0, 100000)
                val sleptNanoSeconds = System.nanoTime() - timeBeforeSleep
                val sleptSeconds = sleptNanoSeconds / 1000000000.0
                secondsLeft -= sleptSeconds.toFloat()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }

    fun cleanup() {}
    abstract fun update(seconds: Float)
}